### ISSUE TITLE
define large DataSet and write the array chunk by chunk after the definition when the data stream flow arrived in the time 

### ISSUE THREAD

**Issue by froglette77950** [jamesmudd/jhdf#654]
To handle very large HDF5 files of several GBytes beyond the capacity of RAM, or to store data as it is created or acquired.

**Comment by jamesmudd** [comment]
Thanks for opening this issue. I agree this would be a nice feature. Just to set expectations, writing support is currently contiguous datasets only (as this is the simplest). So first would be to add any kind of chunked dataset support, initially I would implement single chunk, as again this is the simplest just opens up the possibility of filters support (e.g. compression, verification). Then this would need to be expanded to multi-chunk layouts (maybe b-tree). Once this is implemented then it might be a next step to allow incremental chunk writing as you have requested. Also while huge datasets are an issue IMO its a small use case. I think most users wanting to write HDF5 from Java already have the data in some JVM object.

So essentially while I agree nice feature, it's some time away in my roadmap. Of course if anyone else would like to work on this I would be happy to review the code.

**Comment by jcschaff** [comment]
Another use case for this, from an application that writes rather than reads.

VCell (https://vcell.org) is an open-source platform for cell-biology
simulation. We've just moved all of our HDF5 *reading* to jhdf and deleted the
legacy native binding we'd been carrying for years — thank you, that was a
pleasant migration. One thing blocks us from dropping the native library
altogether: exporting spatial simulation results.

The export creates a dataset of the run's full extent, then streams into it one
time-slice per write:

      dims = { sizeX, sizeY, numSlices, numTimepoints }
      H5Sselect_hyperslab(space, SET, {0, 0, slice, t}, null, {sizeX, sizeY, 1, 1}, null);
      H5Dwrite_double(dataset, ..., sliceData);

so peak memory is one slice regardless of the export's size. With jhdf we'd have
to assemble the whole thing first: a 200×200 grid over 50 slices and 200
timepoints is 3.2 GB, and users request exports that size from a dialog. It's
per-file rather than per-dataset, too — since the file is serialised on close(),
splitting into several datasets wouldn't reduce the peak.

On the roadmap you sketched above: chunked writing landed in #841 and shipped in
0.13.0 (which we're on), so incremental chunk writes look like the immediate
next step rather than a distant one. If you'd still rather not spend time on it,
we'd be glad to attempt a PR — and we can test it against real multi-gigabyte
exports, which may be useful either way.

**Comment by jamesmudd** [comment]
I agree incremental chunk writing now looks closer than ever. With AI you can probably get a PR together and test it out. 
Realistically I won't have time to do this myself but would be very happy to review if you make progress.

**Comment by jcschaff** [comment]
Thank you for the invitation to review the PR, I recognize that this takes time and care.

I would be glad to put a PR together (with some help from AI) and will do a manual review myself.

**Comment by jcschaff** [comment]
Status update, since there is now something to look at.

**#865 is open** and implements the first half: a `ChunkProvider` supplying one chunk at a time, so a dataset larger than memory can be written. It turned out to be a small change — `writeChunkedData` already wrote each chunk straight to the channel and kept only its address, size and offset, and `write` already emitted a placeholder data layout message and rewrote the object header afterwards. The only thing forcing the whole dataset into memory was one line encoding it into a single buffer, plus the `Integer.MAX_VALUE` guard that existed because of it.

Evidence it does what it claims: for a dataset small enough to write both ways, the file produced from a `ChunkProvider` is **byte for byte identical** to the file produced from the whole array — including partial edge chunks and shuffle+deflate. And a test writes **3.2 GB through a 512 MB heap** and reads a slice back (opt-in, since it is slow and writes a multi gigabyte file).

**The second half is written too** — a `StreamingDataset` where the application pushes chunks as data arrives (`writeChunk(offset, data)`), which is closer to what this issue actually asked for. It is held back as a separate PR so #865 can be reviewed on its own; stacking them would make one large diff out of two small ones. It needed the root group object header address to become a superblock field rather than a constant, so streamed data can occupy the front of the file while the tree describing it is still being built. Files that stream nothing are laid out exactly as before.

Both halves have been exercised by a real consumer rather than only by their own tests: VCell's spatial export now writes through them, and its last native HDF5 dependency is gone as a result. Porting it is also what found a gap — `newStreamingDataset` was only available on the file's root group, and that export writes two groups deep.

@jamesmudd whenever you have time to review #865, it would be much appreciated. One thing worth knowing first: **its CI has never run** — the workflows are held at `action_required`, which GitHub does for a fork PR from a first-time contributor, so there is no build attached to it. Approving the run should be all that is needed. In the meantime I ran your CI's own commands (`./gradlew build`, then `jar sourcesJar javadocJar`) on JDK 17, 21, 23 and 25 on macOS arm64: all pass, 1440 tests, and the 14 `h5dump` cross-check tests ran rather than being skipped, so the files are read back by the real HDF5 library. Windows and Linux are the gap your matrix would close.

**Comment by jamesmudd** [comment]
This looks like amazing progress. Thanks for looking into this and for testing it in a real use case. I have run the CI builds which all look good. I have read the code very quickly, but would like to take a better look and consider the API (although I kind of think it might be better to add this and add/improve the API if needed when I know more about other use cases). I'm traveling for the next 2 weeks but will still try and take a look if i get a few hours.

**Comment by jcschaff** [comment]
Thanks so much for agreeing to review my proposal.  This has come together very quickly, I don't want to rush you, it takes longer to review than to create these days.  I broke my contribution into 2 PRs to make it easier to describe and review - please see a draft of my follow-up PR https://github.com/jcschaff/jhdf/pull/2 which has already been integrated and tested in VCell using jitpack.io (see https://github.com/virtualcell/vcell/pull/2001) and is working nicely.

My intention is to support your integration of this work into jhdf repo and return VCell to the official jhdf when this work is complete.

jhdf is a fantastic library - I look forward to working with you on this feature.
